File: chip8_python/cpu.py
from typing import Tuple
import random
import pygame

from memory import Memory
from renderer import Renderer
from sound import Sound
from clock import Clock

"""
A class for handling the CPU operations for the CHIP-8
"""
class CPU:
    def __init__(self, memory: Memory, renderer: Renderer, sound: Sound) -> None:
        self._pc = 0x200 # program counter register, program always loads in at
                         # address 200
        self.i_register = 0x0 # the index register 
        self.registers = [0x0] * 16

        self._delay_timer = 0
        # create pointers to the rest of the hardware components
        self.memory = memory
        self.renderer = renderer 
        self.sound = sound

    def cycle(self) -> None:
        """
        The main fetch-decode-execute cycle of the CPU
        """
        # fetch the instruction at the current PC
        instruction = self.__fetch()
        # decode and execute the instruction
        self.__decode_execute(instruction)

    def decrement_timer(self) -> None:
        """
        Decrement the delay timer
        """
        self._delay_timer -= 1

    def __fetch(self) -> int:
        # chip 8 instuctions are two bytes long
        instruction_byte1 = self.memory.get_from_memory(self._pc)
        instruction_byte2 = self.memory.get_from_memory(self._pc + 1)

        # increment the pc by two bytes 
        self._pc += 2

        # get combine the two 1 byte instructions into one 2 byte instruction
        # concatenate the two hexadecimal bytes into one instruction string
        hex_val = "0x" + format(instruction_byte1, '02x') + format(instruction_byte2, '02x') 

        # return the 2 byte instruction as a hex value (int)
        return int(hex_val, 16) 


    def __decode_execute(self, instruction: int) -> None:
        # first, divide the instruction into bigger categories based on 
        # the first value in the instruction
        match instruction & 0xf000:
            case 0x0000:
                match instruction:
                    case 0x00E0:
                        # clear the screen
                        self.renderer.clear_screen() 
                    case 0x00EE:
                        # return from subroutine by getting the saved pc from the stack
                        self._pc = self.memory.stack.pop() 
            case 0x1000:
                # jump instruction, set the program counter to final three 
                # numbers of hex
                self._pc = instruction & 0x0fff 
            case 0x2000:
                # call to a subroutine - first add the pc to the stack, then set pc
                self.memory.stack.append(self._pc)
                self._pc = instruction & 0x0fff
            case 0x3000:
                # skip one 2 byte instruction if vx is equal to NN in 3XNN
                val = instruction & 0x00ff 
                vx = self.registers[(instruction & 0x0f00) >> 8]

                if val == vx:
                    self._pc += 2
            case 0x4000:
                # skip one 2 byte instruction if vx is NOT equal to NN in 4XNN 
                val = instruction & 0x00ff
                vx = self.registers[(instruction & 0x0f00) >> 8]

                if val != vx:
                    self._pc += 2
            case 0x5000:
                # skip one 2 byte instruction if VX and VY are equal in 5XY0 
                vx = self.registers[(instruction & 0x0f00) >> 8]
                vy = self.registers[(instruction & 0x00f0) >> 4]

                if vx == vy:
                    self._pc += 2
            case 0x6000:
                # set register VX to NN where instruction in form 0x6XNN 
                x = (instruction & 0x0f00) >> 8
                self.registers[x] = instruction & 0x00ff
            case 0x7000:
                # add value to register VX, where instruction in form 0x7XNN 
                x = (instruction & 0x0f00) >> 8
                self.registers[x] = (self.registers[x] + (instruction & 0x00ff)) % 256
            case 0x8000:
                # the 8XYn instructions are logical + arithmetic instructions
                # decoded based on the last 4 bits
                
                vx = self.registers[(instruction & 0x0f00) >> 8]
                vy = self.registers[(instruction & 0x00f0) >> 4]

                match instruction & 0x000f:
                    case 0x0000:
                        # set VX to VY
                        self.registers[(instruction & 0x0f00) >> 8] = vy
                    case 0x0001:
                        # vx = vx | vy
                        self.registers[(instruction & 0x0f00) >> 8] = vx | vy
                    case 0x0002:
                        # vx = vx & vy
                        self.registers[(instruction & 0x0f00) >> 8] = vx & vy
                    case 0x0003:
                        # vx = vx xor vy
                        self.registers[(instruction & 0x0f00) >> 8] = vx ^ vy
                    case 0x0004:
                        # vx = vx + vy (with overflow detection)
                        total = vx + vy

                        if total > 255:
                            # set vf to 1
                            self.registers[0xf] = 1
                        else:
                            self.registers[0xf] = 0

                        self.registers[(instruction & 0x0f00) >> 8] = (vx + vy) % 256
                    case 0x0005:
                        # vx = vx - vy 
                        self.registers[(instruction & 0x0f00) >> 8] = (vx - vy) % 256

                        # if we underflow, then set the VF register to 0
                        if vx > vy:
                            self.registers[0xf] = 1
                        else:
                            self.registers[0xf] = 0

                    case 0x0006:
                        # shift one bit to the right
                        # furthermore, implements original CHIP-8 for COSMIC VIP instruction, first
                        # setting VX to the value at VY
                        
                        # get the bit to be shifted out of vx (last bit)
                        if vy % 2 == 0:
                            # then last bit is 0
                            self.registers[0xf] = 0
                        else:
                            # then last bit is 1
                            self.registers[0xf] = 1

                        # shift vx (= vy) to the right
                        self.registers[(instruction & 0x0f00) >> 8] = vy >> 1

                    case 0x0007:
                        # vx = vy - vx 
                        self.registers[(instruction & 0x0f00) >> 8] = (vy - vx) % 256

                        # if we underflow, then set the VF register to 0
                        if vy > vx:
                            self.registers[0xf] = 1
                        else:
                            self.registers[0xf] = 0
                    case 0x000e:
                        # shift one bit to the left  
                        # furthermore, implements original CHIP-8 for COSMIC VIP instruction, first
                        # setting VX to the value at VY

                        # get the bit to be shifted out of vx (first bit), and we know this register is
                        # 8 bits
                        if (vy & 0x80) >> 7 == 0:
                            # then last bit is 0
                            self.registers[0xf] = 0
                        else:
                            # then last bit is 1
                            self.registers[0xf] = 1

                        # shift vx (= vy) to the right (crop to 8 bits)
                        self.registers[(instruction & 0x0f00) >> 8] = (vy << 1) & 0xff
            case 0x9000:
                # skip one 2 byte instruction if VX and VY are NOT equal in 9XY0 
                vx = self.registers[(instruction & 0x0f00) >> 8]
                vy = self.registers[(instruction & 0x00f0) >> 4]

                if vx != vy:
                    self._pc += 2
            case 0xa000:
                # set the index register to NNN, where instruction is 0xANNN  
                self.i_register = instruction & 0x0fff
            case 0xb000:
                # instruction BNNN - jump with offset, jump to address NNN with offset stored in v0 
                self._pc = self.registers[0] + instruction & 0x0fff
            case 0xc000:
                # generate a random number and bitwise and it with NN 
                rand_val = (instruction & 0x00ff) & (random.random() * (instruction & 0x00ff))
                self.registers[(instruction & 0x0f00) >> 8] = rand_val
            case 0xd000:
                # set the display at coordinates X and Y, instruction is DXYN
                # x and y coords from the registers, modulo the size of 
                # the screen
                vx = self.registers[(instruction & 0x0f00) >> 8] % 64 
                vy = self.registers[(instruction & 0x00f0) >> 4] % 32 
                
                # set vf to 0
                self.registers[0xf] = 0x0

                # the sprite to be drawn to the display has N bytes of data
                n = instruction & 0x000f
                
                # the length of a row is 8, see if this row goes off the right
                # edge of the screen, and if it does, cap it
                row_len = 8
                if vx + 8 >= 64:
                    row_len = 64 - vx

                # the number of rows is n, see if the y coord plus this value
                # exceeds the bottom row of the screen, cap
                col_len = n
                if vy + n >= 32:
                    col_len = 32 - vy


                for offset in range(col_len):
                    # memory address of Nth byte of sprite data
                    memory_address = self.i_register + offset 
                    sprite_byte = self.memory.get_from_memory(memory_address) 
                    sprite_byte = format(sprite_byte, '08b') 

                    # increment through all of the bits of this byte of sprite 
                    # data
                    for bit in range(row_len): 
                        sprite_bit = sprite_byte[bit]

                        y_coord = vy + offset # y coordinate on screen
                        x_coord = vx + bit # x coordinate on screen

                        if sprite_bit == '1':
                            # if the pixel at that coordinate is also on, turn
                            # the pixel off
                            if self.renderer.get_color_at_pixel(row=y_coord, col=x_coord):
                                self.renderer.set_color_at_pixel(row=y_coord, col=x_coord, val=False)
                                # set vf to 1
                                self.registers[0xf] = 1 
                            else:
                                # the pixel is off, so turn it on, keep vf at 0
                                self.renderer.set_color_at_pixel(row=y_coord, col=x_coord, val=True)
            case 0xe000:
                # e based instructions are skip-if key is being pressed
                # first, get the hexadecimal code for the key being stored in vx, 0-f
                instruction_key = (instruction & 0x0f00) >> 8 
                for event in pygame.event.get():
                    if event.type == pygame.KEYDOWN:
                        # get the key that the hexidecimal key corresponds to
                        if (instruction_key in self.memory.valid_keys):
                            # check the key being pressed is the key that vx holds
                            match instruction & 0x000f:
                                case 0x000e:
                                    # instruction skips if key is equal to the instruction key
                                    if event.key == self.memory.valid_keys[instruction_key]: 
                                        # skip the 2 byte instruction
                                        self._pc += 2 
                                case 0x0001:
                                    # instruction skips if key is NOT equal to the instruction key
                                    if event.key != self.memory.valid_keys[instruction_key]: 
                                        # skip the 2 byte instruction
                                        self._pc += 2 
            case 0xf000:
                
                vx = self.registers[(instruction & 0x0f00) >> 8]

                match instruction & 0x00ff:
                    case 0x0007:
                        # set vx register equal to the delay timer
                        self.registers[(instruction & 0x0f00) >> 8] = self._delay_timer
                    case 0x0015:
                        # set the delay timer to value in register
                        self._delay_timer = vx 
                    case 0x0018:
                        # set the sound timer to value in register
                        self.sound.set_timer(vx) 
                    case 0x001e:
                        # add vx to the index register (check for overflow from 0x0fff to 0x1000, set vf accordingly)
                        if (self.i_register + vx > 0x0fff):
                            self.registers[0xf] = 1
                        else:
                            self.register[0xf] = 0
                        self.i_register += vx
                    case 0x000a:
                        # a blocking instruction until a key is pressed.
                        for event in pygame.event.get():
                            if event.type == pygame.QUIT:
                                pygame.quit()
                                exit(0)
                            elif event.type == pygame.KEYUP:
                                # get the key that the hexidecimal key corresponds to
                                if (event.key in self.memory.key_hex):
                                    # set vx to the key being held
                                    self.registers[(instruction & 0x0f00) >> 8] = self.memory.key_hex[event.key] 
                                    # do not execute the else statement
                                    break
                        else:
                            # go back to this same instruction (block)
                            self._pc -= 2

                    case 0x0029:
                        # set index register to location of font of hexadecimal character in memory
                        # each font character is 5 bytes ()
                        self.i_register = self.memory.get_from_memory(5 * vx)

                    case 0x0033:
                        # binary coded decimal conversion (convert the number stored in vx to 3
                        # decimal digits)
                        self.memory.set_memory(loc=self.i_register + 0, val=(vx // 100) % 10)
                        self.memory.set_memory(loc=self.i_register + 1, val=(vx // 10) % 10)
                        self.memory.set_memory(loc=self.i_register + 2, val=(vx // 1) % 10)

                    case 0x0055:
                        registers = self.registers[0: ((instruction & 0x0f00) >> 8) + 1] 
                        for i, register in enumerate(registers):
                            self.memory.set_memory(loc=self.i_register + i, val=register)

                    case 0x0065:
                        x = (instruction & 0x0f00) >> 8
                        for i in range(x + 1):
                            self.registers[i] = self.memory.get_from_memory(loc=self.i_register + i)


                        

File: chip8_cpp/src/cpu.cpp
#include <SDL_events.h>
#include <SDL_keyboard.h>
#include <SDL_scancode.h>
#include <climits>
#include <cpu.h>
#include <cstdint>
#include <cstdlib>
#include <stack>
#include <SDL2/SDL.h>
    
#include <memory.h>
#include <renderer.h>

CPU::CPU(Memory* chip8_memory, Renderer* chip8_renderer) {
    pc_ = 0x200; // start the program counter at the beginning of the loaded ROM
    i_register_ = 0x0;

    CPU::memory_ = chip8_memory;
    CPU::renderer_ = chip8_renderer; 
}

void CPU::decrement_timer() {
    if (delay_timer_ != 0) {
        delay_timer_--;
    }
}

void CPU::cycle() {
    // first get the instruction
    uint16_t instruction = fetch();
    // using the fetched instruction, run the proper function from linked hardware
    decode_execute(instruction);
}

uint16_t CPU::fetch() {
    // read the instruction currently being pointed at by pc
    uint8_t instruction_byte1 = memory_->get_from_memory(pc_);
    uint8_t instruction_byte2 = memory_->get_from_memory(pc_ + 1);
    uint16_t instruction = (instruction_byte1 << 8) + (instruction_byte2);
    // after reading the instructions, increment the pc immediately
    pc_ += 2;
    return instruction; 
}

void CPU::decode_execute(uint16_t instruction) {
    switch (instruction & 0xf000) {
        case 0x0000:
            if (instruction == 0x00e0) {
                // 00e0: clear the screen
                renderer_->clear_screen();
            }
            else if (instruction == 0x00ee) {
                // return from subroutine function
                pc_ = stack_.top();
                stack_.pop();
            }
            break; 
        case 0x1000:
            // jump instruction to set PC to final bytes of hex 
            pc_ = instruction & 0x0fff;
            break; 
        case 0x2000:
            // call subroutine at address NNN from instruction 2NNN
            // push the current pc to the stack so that we can return later
            stack_.push(pc_);
            pc_ = instruction & 0x0fff;
            break; 
        case 0x3000:
            // skip instruction if val in register VX is equal to NN
            {
                uint8_t vx = var_registers_[(instruction & 0x0f00) >> 8];
                if (vx == (instruction & 0x00ff)) {
                    pc_ += 2;
                }
                break; 
            } 
        case 0x4000:
            // skip instruction if val in register VX is not equal to NN 
            {
                uint8_t vx = var_registers_[(instruction & 0x0f00) >> 8];
                if (vx != (instruction & 0x00ff)) {
                    pc_ += 2;
                }
                break; 
            }
        case 0x5000:
            {
                uint8_t vx = var_registers_[(instruction & 0x0f00) >> 8];
                uint8_t vy = var_registers_[(instruction & 0x00f0) >> 4];
                if (vx == vy) {
                    pc_ += 2;
                }
                break; 
            }
        case 0x6000:
            // set register VX to NN, from instruction 6XNN
            var_registers_[(instruction & 0x0f00) >> 8] = instruction & 0x00ff;
            break; 
        case 0x7000:
            // add NN to value in register VX, from instruction 7XNN
            // without setting the carry flag
            var_registers_[(instruction & 0x0f00) >> 8] += instruction & 0x00ff;
            break; 
        case 0x8000:
            {
                uint8_t vx = var_registers_[(instruction & 0x0f00) >> 8]; 
                uint8_t vy = var_registers_[(instruction & 0x00f0) >> 4]; 

                switch (instruction & 0x000f) {
                    case 0x0:
                        // set vx = vy
                        var_registers_[(instruction & 0x0f00) >> 8] = var_registers_[(instruction & 0x00f0) >> 4];
                        break;
                    case 0x1:
                        // set vx to the bitwise OR of vx and vy
                        var_registers_[(instruction & 0x0f00) >> 8] = vx | vy;
                        break;
                    case 0x2:
                        // bitwise AND
                        var_registers_[(instruction & 0x0f00) >> 8] = vx & vy;
                        break;
                    case 0x3:
                        // bitwise XOR 
                        var_registers_[(instruction & 0x0f00) >> 8] = vx ^ vy;
                        break;
                    case 0x4:
                        {
                            // ADD VX and VY and affect the overflow flag
                            uint16_t total = vx + vy;
                            // if there is an overflow, set the flag
                            if (total > 255) {
                                var_registers_[0xf] = 1;                            
                            }
                            else {
                                var_registers_[0xf] = 0;
                            }
                            var_registers_[(instruction & 0x0f00) >> 8] = static_cast<uint8_t>(vx + vy);
                            break;
                        }
                    case 0x5:
                        // set the carry flag to 0 if we underflow
                        if (vx > vy) {
                            var_registers_[0xf] = 0;
                        }
                        else {
                            var_registers_[0xf] = 1;
                        }
                        // subtract: vx = vx - vy
                        var_registers_[(instruction & 0x0f00) >> 8] = vx - vy;
                        break;
                    case 0x6:
                        // shift vx one bit to the right and set carry flag appropriately
                        if (vy % 2 == 0) {
                            // last bit is zero
                            var_registers_[0xf] = 0;
                        }
                        else {
                            var_registers_[0xf] = 1;
                        }

                        var_registers_[(instruction & 0x0f00) >> 8] = vy >> 1;

                        break;
                    case 0x7:
                        // set the carry flag to 0 if we underflow
                        if (vy > vx) {
                            var_registers_[0xf] = 0;
                        }
                        else {
                            var_registers_[0xf] = 1;
                        }
                        // subtract: vx = vx - vy
                        var_registers_[(instruction & 0x0f00) >> 8] = vy - vx;
                        break;
                    case 0xe:
                        if ((vy & 0x80) >> 7 == 0) {
                            // last bit is 0
                            var_registers_[0xf] = 0;
                        }
                        else {
                            var_registers_[0xf] = 1;
                        }

                        // shift vx (= vy) to the right (crop to 8 bits)
                        var_registers_[(instruction & 0x0f00) >> 8] = (vy << 1);
                        break;
                } 
            }
            break; 
        case 0x9000:
            {
                uint8_t vx = var_registers_[(instruction & 0x0f00) >> 8];
                uint8_t vy = var_registers_[(instruction & 0x00f0) >> 4];
                if (vx != vy) {
                    pc_ += 2;
                }
                break; 
            }
        case 0xa000:
            // set index register
            i_register_ = instruction & 0x0fff;
            break; 
        case 0xb000:
            // jump with NNN + offset stored in v0
            pc_ = var_registers_[0x0] + (instruction & 0x0fff);
            break; 
        case 0xc000:
            // vx = rand & NN
            var_registers_[(instruction & 0x0f00) >> 8] = (instruction & 0x00ff) & static_cast<uint8_t>((rand() % instruction & 0x00ff));
            break; 
        case 0xd000:
            {
                // draw to display instruction
                // get the X and Y coordinates from the VX and VY registers
                uint8_t vx = var_registers_[(instruction & 0x0f00) >> 8] % SCREEN_WIDTH; 
                uint8_t vy = var_registers_[(instruction & 0x00f0) >> 4] % SCREEN_HEIGHT; 

                // set the VF register to 0
                var_registers_[0xf] = 0x0;

                // cap the columns/rows to be drawn if the rows/cols exceed the screen size
                uint8_t row_len = 8;
                if (row_len + vx >= SCREEN_WIDTH) {
                    row_len = SCREEN_WIDTH - vx;
                }

                // sprite to be drawn has N bytes of data col_len = N
                uint8_t col_len = instruction & 0x000f;
                if (col_len + vy >= SCREEN_HEIGHT) {
                    col_len = SCREEN_HEIGHT - vy;
                }

                // for every possible column val (every byte of the sprite)
                for (int offset = 0; offset < col_len; offset++){
                    uint8_t sprite_byte = memory_->get_from_memory(i_register_ + offset);
                    // increment through each bit of byte of sprite data
                    for (int bit = 0; bit < row_len; bit++) {
                        // get one bit at a time, starting from the MSB
                        uint8_t sprite_bit = (sprite_byte & (1 << (7 - bit))) >> (7 - bit); 
                        uint8_t y = vy + offset; // y coordinate on display
                        uint8_t x = vx + bit; // x coordinate on display

                        if (sprite_bit == 1) {
                            if (renderer_->get_pixel_is_on(x, y)) {
                                renderer_->set_pixel(x, y, false);
                                var_registers_[0xf] = 1;
                            }
                            else {
                                renderer_->set_pixel(x, y, true);
                            }
                        }
                    }
                }
                break; 
            }
        case 0xe000:
            {
                // check if a key is being pressed, and if it is
                // if the key that is being pressed is a valid key in the CHIP 8 system 
                const uint8_t* state = SDL_GetKeyboardState(nullptr);
                uint8_t scancode = memory_->hex_to_scancode_translation[(instruction & 0x0f00) >> 8];
                if (((instruction & 0x000f) == 0xe) && state[scancode]) {
                    // the key being queried is being pressed
                    pc_ += 2;
                } else if (((instruction & 0x000f) == 0x1) && !state[scancode]) {
                    // TODO: this checks if the key is not being pressed, but not if it is a valid key on the CHIP-8 system
                    pc_ += 2;
                }
                break;
            }
        case 0xf000:
            switch (instruction & 0x00ff) {
                case 0x07:
                    // set the delay timer from the VX register
                    delay_timer_ = var_registers_[(instruction & 0x0f00) >> 8];
                    break;
                case 0x15:
                    // get the delay tier into the VX register
                    var_registers_[(instruction & 0x0f00) >> 8] = delay_timer_;
                    break;
                case 0x18:
                    // TODO : set the sound timer 
                    break;
                case 0x1e:
                    // add value in vx to index register
                    i_register_ += var_registers_[(instruction & 0x0f00) >> 8];
                    break;
                case 0x0a:
                    // get key (blocking call)
                    {
                        bool key_up = false;
                        SDL_Event event;
                        while (SDL_PollEvent(&event)) {
                            switch (event.type) {
                                case SDL_QUIT:
                                    renderer_->quit();
                                    exit(0);
                                case SDL_KEYUP:
                                    uint8_t key_released_scancode = event.key.keysym.scancode;
                                    if (memory_->scancode_to_hex_translation.find(key_released_scancode) != memory_->scancode_to_hex_translation.end()) {
                                        // set register VX to the hex translation
                                        var_registers_[0xf] = memory_->scancode_to_hex_translation[key_released_scancode];
                                        key_up = true;
                                        break;
                                    }
                            }
                        }

                        // made it through the event loop without triggering the SDL_KEYUP event, so made a blocking call
                        if (!key_up) {
                            pc_ -= 2;
                        }
                    }
                    break;
                case 0x29:
                    // instruction : load font character
                    // each font character is 5 bytes long, so take vx * 5 to find the appropriate character 
                    i_register_ = memory_->get_from_memory(5 * var_registers_[(instruction & 0x0f00) >> 8]);
                    break;
                case 0x33:
                    // binary-coded decimal conversion (extract digits)
                    {
                        uint8_t vx = var_registers_[(instruction & 0x0f00) >> 8];
                        memory_->set_memory(i_register_ + 0, static_cast<uint8_t>(vx / 100) % 10);
                        memory_->set_memory(i_register_ + 1, static_cast<uint8_t>(vx / 10) % 10);
                        memory_->set_memory(i_register_ + 2, static_cast<uint8_t>(vx / 1) % 10);
                    }
                    break;
                case 0x55:
                    // store registers in memory up to register VX (inclusive) 
                    {
                        uint8_t x = (instruction & 0x0f00) >> 8;
                        for (uint8_t i = 0; i <= x; i++) {
                            uint8_t vx = var_registers_[(instruction & 0x0f00) >> 8];
                            memory_->set_memory(i_register_ + i, vx);
                        }
                    }
                    break;
                case 0x65:
                    // load memory into registers
                    {
                        uint8_t x = (instruction & 0x0f00) >> 8;
                        for (uint8_t i = 0; i <= x; i++) {
                            var_registers_[i] = memory_->get_from_memory(i_register_ + i);
                        }
                    }
                    break;
            }
            break;
    }
}
